presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java

```diff
@@ -139,6 +139,23 @@
         };
     }
 
+    @Test
+    public void testPossibleExponentialBacktracking()
+    {
+        testStatement("SELECT CASE WHEN " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * " +
+                        "1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9",
+                "line 1:375: mismatched input '<EOF>'. Expecting: '%', '*', '+', '-', '/', 'AT', 'THEN', '||'");
```

Comment:
i'd expect `IS` too, for example in `CASE WHEN {multiplication} IS NULL THEN 'zonk' ELSE ..`

(just noting, no change requested)

Reply:
That's partly because our grammar does: `CASE <valueExpression> WHEN ...`. Looking at the spec, it seems to allow more complex expressions, including our `<booleanExpression>` (I'll fix that as a separate change). Even them, it suffers from the same problem I explained regarding `SELECT ... FROM ... WHERE a.`